### DISCUSSION TITLE
一个正则表达式引发的惨案

### DISCUSSION THREAD

**Discussion by jiegec** [physics-data/faq#106] [Ideas] [answered]
在 @zzj-zzj-zzj 同学的帮助下，助教发现了一个神奇的现象：

```shell
[[ 1 =~ [+-\\*/] ]] && echo "true"
```

在 bash 中运行上面这一条命令，有的系统上什么也不输出，有的系统上会输出 true。这是为什么呢？

如果你想直接知道答案，请直接跳到最后。

接下来，讲一讲助教是怎么定位到问题的。首先，助教在不同的系统中做了测试，通过控制变量法，排除了 OS 的影响、bash 版本的影响和 bash shopt 设置的影响等等，甚至还去 bash 的源代码树中去看历史记录，对比不同版本历史的更新；也考虑了是不是正则表达式库的实现问题，但都一一排除了。

想来想去，我们对比了不同环境下的环境变量，发现有一个地方可能会导致问题：当 LANG 环境变量设为 `en_US.UTF-8` 时，上面的命令不会输出；当 LANG 环境变量设为 `""` 空时，上面的命令会输出 ` true`。这就很奇怪了，我写一个正则表达式，和你选择什么语言有关系吗？况且，我这个正则表达式是要匹配 `+ - * /` 这四个字符之一，和语言也没有关系啊。

等等。忽然发现一个问题，就是 - 号在语法中表示一段连续的字符，比如 `a-z` 表示了小写字母，那会不会是上面的程序实际上对应了 `+` 号到 `*` 号之间的所有字符，但是，这怎么会匹配到 1 呢。

于是上网查了一下，关键词：regex range order lang bash，找到了文章：<https://stackoverflow.com/questions/43448655/weird-behavior-of-bash-glob-regex-ranges> 。它提到这么一个情况，就是当 `LANG=en_US` 的时候，排序的顺序是 `aAbB`，而不是 `abAB`。于是，接着做了下面的实验：

```shell
$ cat poc.txt | tr '\\n' ' '
1 + - * / \ a b A B 一 二 测 试 α
$ LANG="" sort poc.txt | tr '\\n' ' '
* + - / 1 A B \ a b α 一 二 测 试
$ LANG="zh_CN.UTF-8" sort poc.txt | tr '\\n' ' '
* + - / \ 1 测 二 试 一 a A b B α
$ LANG="en_US.UTF-8" sort poc.txt | tr '\\n' ' '
* + - / \ 1 a A b B α 一 二 测 试
```

首先可以发现是，不同的 locale 下，排序的结果很不一样，但还是不能解释遇到的问题，毕竟 `*` 在最前面。回到一开头的表达式，仔细一看，哦，原来只有一次转义，意味着表达式是从 + 到 \ ，正好，当 `LANG=""` 时，1 夹在这两个字符之间，所以匹配为 `true`。

**结论** ：不同的 locale 下，字符的排序不同。解决方法：

1. 使用时覆盖掉语言相关的环境变量（`LC_LOCALE` 等）
2. 谨慎使用取范围的语法

**Comment by heroxbd** [comment, marked as answer]
把 `-` 放在最后可以避免歧义。

```
$ [[ - =~ [*+/-] ]] && echo yes
yes
```

`[[ ]]` 之内不会有其它转义，见 `bash(1)`

> Word splitting and pathname expansion are not performed on the words between the [[ and ]]; tilde expansion, parameter and variable expansion, arithmetic expansion,  command  substitution,  process substitution, and quote removal are performed.
